### DISCUSSION TITLE
Handling hours for the numerical habits

### DISCUSSION THREAD

**Discussion by Stiralbios** [iSoron/uhabits#953] [Feature Requests]
Numerical habbits have subunit but the amount of subunits is always 100.
For example you can set a tracker 2.80. If you want to track a time in hours it's not very practical (2.80 hours = 2h48min).
It could be usefull to be able to personnalize the amount of subnits, this would allow the users to represent the hours (and others units) in an easier way to understand

**Comment by azxcqer** [comment]
maybe it could have a dedicated "hours" type of units so that would be an option because I can't seen to think of any other subunits that would need to be customizable but that is just me.

or maybe it simple math could be added to the numerical field so you could quickly calculate the number ( x/60 will give you the result in the 1-100 scale)

**Reply by Stiralbios** [reply]
Americains have weird length units. You can be 5"11 tall for example but there is no 5"12, if you are a bit taller you are 6"
But yes I am not aware of a lot of unit that have this propertie. The main ones are time units. And the hours is likelly the most usefull one.

**Reply by herrbrixter** [reply]
Time should be a dedicated unit. For example,

Work duration: 14 hours and 47 minutes or 14:47

People can also use it for time of the day as long as they treat it as military time

10:15PM is equivalent to 22:15

**Comment by DeWarner** [comment]
i think this numerical selector is not ideal for time, a clock type selector. e.g. https://www.tutlane.com/tutorial/android/android-timepicker-with-examples
might be better

**Comment by jrbergen** [comment]
Would a feature request for a time-based goal fit under thisone? For instance a habit to go sleep, meditate, or wake up at a specific time.

Currently one could use numbers 0 to 2359 as a workaround, but this is a bit awkard as you'd get a negative or positive score for being too early or too late. An absolute distance w.r.t. a specificied goal time might be more applicable (with e.g. a delta of < -12 or > 12 hours from the target counting it for the previous and next day, respectively).

Currently I don't think I have the knowledge to try and implement it myself. However I'm picking up Kotlin and Android development, so hopefully I'll be able to help somewhere in 2022 if maintainers agree this feature would be worth it to add.

**Reply by daisystanton** [reply]
+1 to this! I want to track my wake, sleep, and "last meal" time, and logging "2115" (for 9:15pm) currently gets rendered as "2.1k", which is understandable by me but not ideal.

**Comment by ke352802081770314** [comment]
personally I've defined time-based goals (like time spent gardening, or walking/cycling) by using minutes as the unit, and I ignore the fractional part, only enter whole numbers. But having a separate "time" datatype would also permit features like a "start now" and "end now" button, calculating the duration automatically, for those users who use the tracker in real-time instead of entering the record later.

**Comment by AleeRezaa** [comment]
I think it would be very helpful to add "time-based" habits beside numerical ones. We can also have stopwatches for every habit of this type. For example I want to study for 2 hours everyday and exercise for 1 hour. So I need two stopwatches and the app can calculate every habit by its stopwatch's time.

**Comment by JosephChataignon** [comment]
I have nothing to add, just adding any form of handling hours would still be very nice in 2025

**Reply by neube3** [reply]
Well, maybe not "any". I, for one, absolutely despise the "select on clock" and "roll the hours" input types. 
Intuitive number input is a very strong and welcome feature: 
• if it's two fields, allow us to move between them with keyboard (think: sliding in space in gboard)
• not requiring us to type in ":" but allowing for it if it's single field
• typing 0 in hours jumping to minutes input  directly
• being smart enough to know that if I type in "257" it means 2h57m and not returning with an error "cannot enter 25h" or similar,
• allowing for typing in >60 minutes and calculating the hours from it (e.g. 123 minutes -> 02h03m) if you start typing in minutes subfield with 0 in hours
• probably more input magic I didn't think about

Oh, and updating this wish for 2026, I guess.

**Reply by herrbrixter** [reply]
I also hate the clock input and swiping to roll a dial. I prefer typing out the number to input the time.

If I want to put 14 hours and 38 minutes, I just type 1 - 4 - 3 - 8.

If I want to put just 27 minutes, I just type 2 - 7.

If I want to out 1 hour and 8 minutes, I just type 1 - 0 - 8.

**Comment by herrbrixter** [comment]
It would be great to be able to input time units as hours:minutes.

In addition, please allow negative time duration similar to #1385. For example, "-0:20" means I did an activity 20 minutes faster than expected.

**Comment by dalewking** [comment]
I too am looking for time based units for things like bed time and up and ready time, but it seems this discussion is confused by people asking for 2 different things but using the word "Time" to represent both concepts. It might be good to differentiate
- Durations, how much time is spent
- Time of day at what time something occurred
